communicate default style safely

M-frankied opened the pull request:
if no explicit style was specified csharp-mode called c-set-style to set style "C#". But this caused a lot of cc-mode code to be executed before initializing cc-mode, causing trouble (see issue #117).

This change achieves the intended effect in a different way. It binds "C#" to c-default-style for the duration of c-common-init if no explicit style was specified and lets nature take its course.

M-frankied commented:
I'm going to go ahead and assume it wasn't my PR that broke the build :-)

josteink commented:
> I'm going to go ahead and assume it wasn't my PR that broke the build :-)

Yeah. I've seen similar issues for other repos I manage, and they've all been for Emacs git master... I'm not happy about it, but haven't yet decided what the appropriate course of action is.

> this caused a lot of cc-mode code to be executed before initializing cc-mode, causing trouble (see issue #117).
>
> This change achieves the intended effect in a different way. It binds "C#" to c-default-style for the duration of `c-common-init` if no explicit style was specified and lets nature take its course.

So basically your solution to this problem is *avoid* calling `c-set-style`, because this call is the source of the problems you encounter?

If so, I would also take care to document this in form of a comment before the `(unless ...)` block, because this kind of documentation needs to be close to the relevant code, so that we don't forget why the code was structured this way in the future, and cause new regressions.

Speaking of regressions... Would it be possible to add a test-case for this? Basically re-bind both `c-set-style` and `c-common-init` before calling `csharp-mode` and verifying that `c-set-style` was not invoked before `c-common-init`? That would be a even better way to avoid regressions.

M-frankied commented:
I added a comment explaining what the code does, I added a tests that shows the bug I fixed, and I added a test that shows style selection still works as before.

josteink commented:
Thanks. Appreciated. It all looks good and I'll try to do some reviewing and testing as soon as possible.

Nevermind the CI. I'll ease up on the Emacs git-master requirements, and we'll get this merged.

josteink commented:
I've tested and this all looks good to me. Doesn't break anything which isn't already broken :)

@wasamasa: Do you see anything obviously wrong with this, or should I just go ahead and merge?

wasamasa commented on csharp-mode-tests.el:
```diff
@@ -397,6 +397,33 @@
+      (csharp-mode))
+    (should
+     (equal before adaptive-fill-regexp))
+  ))
```
Please don't spill the pringles.

wasamasa commented on csharp-mode-tests.el:
```diff
@@ -397,6 +397,33 @@
+      (csharp-mode)
+      (should
+       (equal "C#" c-indentation-style))
+      )))
```
Same as above.

wasamasa commented:
Looks OK to me aside from minor style issues.

josteink commented:
Looking good. Thanks for the PR :)
